src/core/pusher.ts

```diff
@@ -116,26 +134,49 @@ export default class Pusher {
           channel.handleEvent(event);
         }
       }
+
       // Emit globally [deprecated]
       if (!internal) {
         this.global_emitter.emit(event.event, event.data);
       }
     });
+
     this.connection.bind('connecting', () => {
       this.channels.disconnect();
     });
+
     this.connection.bind('disconnected', () => {
       this.channels.disconnect();
     });
+
     this.connection.bind('error', err => {
       Logger.warn(err);
     });
+  }
 
-    Pusher.instances.push(this);
-    this.timeline.info({ instances: Pusher.instances.length });
+  /**
+   * Allows you to switch Pusher cluster without
+   * losing all the channels/subscription binding
+   * as this is internally managed by the SDK.
+   */
+  switchCluster(options: ClusterOptions) {
+    // Automatically disconnect the current connection
+    this.disconnect();
+
+    const { appKey, cluster } = options;
+
+    this.key = appKey;
+    this.options = { ...this.options, cluster };
+    this.config = getConfig(this.options, this);
+
+    // Recreate a new timeline instances and connection
+    // for the new tuple (appKey, cluster)
+    this.setupTimeline();
+    this.setupConnection();
```

Comment:
At this point, some listeners are still bound to the old connection. All listeners on the old connection needs to be unbound to:
- Alleviate any side effects
- Prevent memory leaks

A side note:
The fact that we will forcefully remove all listeners from the old connection here reveals the danger of implementing `switchCluster` in this way because how would you know which objects/actors need such listeners in order to ask them to bind to reinitialize themselves and bind to the new connection? Examples include `UserFacade` which we know we need to reinitialize just because we looked at it's code and found that it binds to the old connection.
The point of this paragraph is to reveal that fact that the code is stateful in terms of binding listeners to the connection. Creating a new connection in `pusher.connnection` is not enough to guarantee consistency.


Given that note:
- It's crucial to have unit tests to test all functionalities after `switchCluster` is called
- I'd rather implement `switchCluster` in a way that maintains the invariants without the need to look inside each object to make sure it still works.
- Or change the code to make it agnostic to the connection change. For example, introducing an intermediate EventEmitter as a connection delegate all objects bind to and whenever there is a new connection, the new connection will be wired with the delegate while the old connection is unwired from the delegate.

Reply:
Thanks for detailed comment.
From my understanding, calling `this.connection.disconnect` will result in closing the WebSocket connection as well as freeing up everything bound to the that connection.

Now, there is a valid concern, as `connection` is exposed to the outside world, we could still have some event listeners that are outside of the Pusher SDK, we could either loop through them and bind them again to the same connection or just "unbind_all" to avoid any memory leaks and move that responsibility to the consumer.

I'd still challenge the reason behind this choice of making internal resources available for external usage, but that's a bit outside of this MR scope.

I'm just afraid that we are going too deep in some assumptions and making this PR handle things which are outside of the scope. would you be conformable having Pusher team working on ensuring that those assumptions are verified and maybe create new PR to complete the "missing parts" of this implementation?

I just came up with an idea, what if, we don't create a new `ConnectionManager` instance, but instead update its internal config and trigger a `reconnect` (like if the network went from offline to online) this way we won't need to unbound any events and everything would work as expected?

What do you think?

Reply:
I think I came up with something even more simpler and would answer your concerns as it uses literally the same technique as when the Network status changes from offline to online. I will commit shortly.

Reply:
That's definitely a better approach. The ConnectionManager is the delegate in this case that redirects signals between the active connection and the application.

I think the new implementation is sound. You may even revert the unnecessay refactoring to make the PR even simpler. I think it's still important to run unit tests after switchCluster to make sure all feature will still work.